tests/test_clean.py

```diff
@@ -1,96 +1,210 @@
-        bleach.clean('another <em>good</em> tag') ==
+    ),
+
+    (
+        'another <em>good</em> tag',
         'another <em>good</em> tag'
     )
+])
+def test_content_has_allowed_html(data, expected):
+    assert clean(data) == expected
 
 
-def test_bad_html():
+def test_html_is_lowercased():
     assert (
-        bleach.clean('a <em>fixed tag') ==
-        'a <em>fixed tag</em>'
+        clean('<A HREF="http://example.com">foo</A>') ==
+        '<a href="http://example.com">foo</a>'
     )
 
 
-def test_function_arguments():
-    TAGS = ['span', 'br']
-    ATTRS = {'span': ['style']}
+@pytest.mark.parametrize('data, should_strip, expected', [
+    # Regular comment
+    (
+        '<!-- this is a comment -->',
+        True,
+        ''
+    ),
 
-    text = 'a <br/><span style="color:red">test</span>'
-    assert (
-        bleach.clean(text, tags=TAGS, attributes=ATTRS) ==
-        'a <br><span style="">test</span>'
+    # Open comment with no close comment bit
+    (
+        '<!-- open comment',
+        True,
+        ''
+    ),
+    (
+        '<!--open comment',
+        True,
+        ''
+    ),
+    (
+        '<!-- open comment',
+        False,
+        '<!-- open comment-->'
+    ),
+    (
+        '<!--open comment',
+        False,
+        '<!--open comment-->'
+    ),
+
+    # Comment with text to the right
+    (
+        '<!-- comment -->text',
+        True,
+        'text'
+    ),
+    (
+        '<!--comment-->text',
+        True,
+        'text'
+    ),
+    (
+        '<!-- comment -->text',
+        False,
+        '<!-- comment -->text'
+    ),
+    (
+        '<!--comment-->text',
+        False,
+        '<!--comment-->text'
+    ),
+
+    # Comment with text to the left
+    (
+        'text<!-- comment -->',
+        True,
+        'text'
+    ),
+    (
+        'text<!--comment-->',
+        True,
+        'text'
+    ),
+    (
+        'text<!-- comment -->',
+        False,
+        'text<!-- comment -->'
+    ),
+    (
+        'text<!--comment-->',
+        False,
+        'text<!--comment-->'
     )
+])
+def test_comments(data, should_strip, expected):
+    assert clean(data, strip_comments=should_strip) == expected
 
 
-def test_named_arguments():
-    ATTRS = {'a': ['rel', 'href']}
+@pytest.mark.parametrize('data, expected', [
+    # Disallowed tag is escaped
+    ('<img src="javascript:alert(\'XSS\');">', '&lt;img src="javascript:alert(\'XSS\');"&gt;'),
+
+    # Test with parens
+    ('a <script>safe()</script> test', 'a &lt;script&gt;safe()&lt;/script&gt; test'),
+
+    # Test with braces
+    ('a <style>body{}</style> test', 'a &lt;style&gt;body{}&lt;/style&gt; test'),
+])
+def test_disallowed_tags(data, expected):
+    assert clean(data) == expected
 
-    text = '<a href="http://xx.com" rel="alternate">xx.com</a>'
-    assert bleach.clean(text) == '<a href="http://xx.com">xx.com</a>'
+
+def test_invalid_char_in_tag():
+    # NOTE(willkg): Two possible outcomes because attrs aren't ordered
```

Comment:
Is there any downside to ordering attrs? I can't think of any getters that depend on their order though looking below it matters for escaped tags.

Reply:
I'm pretty sure the ordering here happens in html5lib. I'll write up an issue to double-check this. It's possible I fixed the issue that required us to test two possibilities.